### DISCUSSION TITLE
Simpson & Partners Charger

### DISCUSSION THREAD

**Discussion by edc1** [lbbrhzn/ocpp#1372] [General]
Anyone else here using this integration with a Simpson & Partners charger? I have one of their Home 7 Plus chargers and I thought I'd share my experience using it.

## Setup

You configure the web socket URL that points the charger to Home Assistant in the S&P app under Settings -> OCPP. There's a bug in the iPhone app where you start typing the URL in the box and the characters disappear after a second or two. The only way I've found around this is to have the URL in your clipboard, paste it in the box, and quickly tap the save and restart button. That will configure the charger to use OCPP. I don't know if this bug is also present in the Android version of the app as I've only tried it on the iPhone one.

## Measurands

The integration successfully detects the following measurands from the charger:

```
Current.Import
Current.Offered
Energy.Active.Import.Interval
Energy.Active.Import.Register
Power.Active.Import
Power.Offered
Temperature
Voltage
```

Unfortunately for the Temperature measurand, the charger reports its unit as '°C' instead of 'Celsius'. That causes schema validation to fail as '°C' isn't part of the OCPP 1.6-J schema specification. I disable automatic measurand detection and manually select all the ones above apart from Temperature. Possibly having both 'Automatic detection of OCPP Measurands' and 'Skip OCPP schema validation' enabled would work, but I think getting the integration to validate the messages from the charger is rather important.

## Reliability

I've not found using OCPP with this charger to be very reliable. The charger often (multiple times per week) reports InternalError and enters a Faulted state:

```
[2,"xxxxxxxxxxxx-22c4f223-00000097","StatusNotification",{"connectorId":1,"status":"Faulted","errorCode":"InternalError"}]
```

This is particularly annoying when you need it to charge the car in the middle of the night and it doesn't. The only way to get it working again is to restart it from the S&P app. Sometimes twice. This seems to happen a lot more frequently since the 2.2.37 firmware was pushed to the charger. I've reported this to Simpson & Partners so we'll see if they take any notice. Disabling OCPP and just relying on a schedule set on the charger is reliable (but obviously no Home Assistant integration then). I suspect it's the OCPP code in the charger's firmware that's unreliable.

Does anyone have a reliable setup with this charger?

**Comment by m01** [comment]
Did you also report the temperature unit issue to S&P?

**Reply by edc1** [reply]
No, I've not reported the Temperature measurand issue to them, but I can do. That doesn't actually prevent the charger from charging.

The Faulted state with InternalError does prevent charging from starting however.

**Reply by m01** [reply]
I think it'd be a good idea to report it, as fixing it would make the charger work "out of the box".

**Reply by edc1** [reply]
I've reported that to them too now. We'll see if they fix it.

**Reply by m01** [reply]
Looks like in the current firmware version the temperature units are fixed! :tada:
```
Nov 23 ... <hostname> hass[...]: 2024-11-23 ... INFO (MainThread) [ocpp] : receive message [2,"...","MeterValues",{"connectorId":1,"transactionId":...,"meterValue":[{"timestamp":"2024-11-23T...",...,{"measurand":"Temperature","value":"28.500","unit":"Celsius"},... ]}]}]
```
I'm no longer seeing the `"fixing"` warnings from my temporary workaround (and indeed, I don't see errors after removing my workaround).

**Reply by edc1** [reply]
Cool. Which version are you running? I must admit I didn't check if this was fixed in 2.2.39.

**Reply by m01** [reply]
2.2.40

**Reply by edc1** [reply]
Good stuff. I'm glad they've fixed that one!

**Comment by drc38** [comment]
You can use the OCPP integration's Get Diagnostics action to retrieve the charger log and this may shed some light on why the internal error is occurring. If you search the Discussions/Issues there is more info on how to do this.

**Reply by edc1** [reply]
Thanks very much for the suggestion. It doesn't look like I can get diagnostics data from the charger. Querying the configuration for ```SupportedFileTransferProtocols``` I get the following back:

```2024-10-30 19:54:28.316 WARNING (MainThread) [custom_components.ocpp] Get Configuration returned unknown key for: SupportedFileTransferProtocols```

When the charger starts, it tells me it supports the following feature profiles:

```
Core
RemoteTrigger
SmartCharging
```

The ocpp.get_diagnostics action is only registered in Home Assistant if the charger also supports:

```
FirmwareManagement
```

Just to see what would happen, I modified the relevant lines in api.py so the ocpp.get_diagnostics action was availalbe in Home Assistant. When I run it the charger reports ```NotImplemented```:

```
2024-10-30 20:04:22.105 INFO (MainThread) [ocpp] : send [2,"bf357f5f-bac4-4aeb-a3bf-18bb1adc4c05","GetDiagnostics",{"location":"https://webhook.site/abc"}]
2024-10-30 20:04:22.306 INFO (MainThread) [ocpp] : receive message [4,"bf357f5f-bac4-4aeb-a3bf-18bb1adc4c05","NotImplemented","",""]

2024-10-30 20:05:34.824 INFO (MainThread) [ocpp] : send [2,"9761f9b3-1d0c-450a-828a-8ebb12826efd","GetDiagnostics",{"location":"ftp://webhook.site/abc"}]
2024-10-30 20:05:34.905 INFO (MainThread) [ocpp] : receive message [4,"9761f9b3-1d0c-450a-828a-8ebb12826efd","NotImplemented","",""]
```

**Comment by m01** [comment]
I can confirm the iOS app quirk with the edit field and the validation error with the temperature units. I worked around the latter by temporarily patching the Python ocpp library. I haven't yet seen the reliability issue - but I still need to finish one leg of the network cabling, so I've only had temporarily-installed network cables in place for a few days at a time.

I was wondering if the validation error would cause wide-spread issues, but another open-source OCPP "server" seems to just ignore anything other than (k)Wh, at least for display (https://github.com/steve-community/steve/issues/1514). I'm not sure how strict they are with validation. That said, I'm happy this one logs the extra data.

**Reply by edc1** [reply]
S&P have identified the cause of the InternalError errors I was getting and pushed a version of the firmware to my charger with a fix. I've not seen the errors since. I expect that fix will be more widely deployed reasonably soon. 

I've been told they will address the Temperature measurand reporting the wrong unit soon too. So hopefully that won't be a problem for much longer. That issue doesn't stop charging from working. 

I've been impressed with the level of support I've received for these two issues.

**Comment by bleara** [comment]
I was reading this post & started to think I had written it because you are seeing exactly the behaviour that I get. I can confirm that the Android App has or at least had the same issue when trying to type in the URL, so having to do a quick copy and paste and save has been the only way to do it. I am told that this is fixed in the latest version of the App but I haven't tried it.
The reliability problems when using OCPP are just as I see too. I had observed this getting worse over the last few releases of the firmware but I had also been adding in code in HA to try and work around the occasional problems These changes I have made seem to make it worse, for instance checking  'charger_status' once a minute had my charger lock up twice in a day on one occasion this week.
S & P have put me on their beta test program so I am currently using 2.2.39 which has recovered from an 'unavailable' state but not always. 
I have found that the more I read the 'charger_status' the worse it seems to get so if you can just do a start charge and stop charge without status checks then it works reliably so far (2 days at least). I am feeding back to S&P with my findings and they do sound keen to fix the issues.
How do you get the log data ? I've not gone to that level in HA so it's new to me and I can't see anything relevant in the system logs.

**Reply by edc1** [reply]
Ah, it's not just me then!

I'm running 2.2.39 on my charger too. It definetly seems more reliable than 2.2.37 and 2.2.38. It does recover sometimes from the InternalError errors. I've seen this a couple of times in the OCPP logs in Home Assistant (running 2.2.39):

```
2024-11-16 18:19:14.179 INFO (MainThread) [ocpp] : receive message [2,"xxxxxxxxxxxx-43bb5692-00000298","StatusNotification",{"connectorId":1,"status":"Faulted","errorCode":"InternalError"}]
2024-11-16 18:19:14.183 INFO (MainThread) [ocpp] : send [3,"xxxxxxxxxxxx-43bb5692-00000298",{}]
2024-11-16 18:19:24.447 INFO (MainThread) [ocpp] : receive message [2,"xxxxxxxxxxxx-540e2e7a-00000299","StatusNotification",{"connectorId":1,"status":"Available","errorCode":"NoError"}]
2024-11-16 18:19:24.450 INFO (MainThread) [ocpp] : send [3,"xxxxxxxxxxxx-540e2e7a-00000299",{}]
```

You can enable debug logging for the OCPP integration in your configuration.yaml in Home Assistant. You can also enable debug logging for the websockets server (need to restart HA after adding those lines):

```
logger:
  default: info
  logs:
    custom_components.ocpp: debug
    websockets.server: debug
```

I've not found the logs from the websockets server to be particularly useful. For these problems at least. So I just have the OCPP debug logs enabled. You can see the logs under Settings -> System -> Logs and filter it for 'ocpp' in the UI. Or in /config/home-assistant.log. From a terminal:

```
grep -i ocpp /config/home-assistant.log | less
```

I have setup some automation in Home Assistant that alerts me when the sensor.charger_status_connector entity goes into Faulted state or becomes unavailable. And something similar on the sensor.charger_version_firmware entitly that alerts me when the firmware version changes.

Interesting that you can get it to fail faster by checking the charger status more frequently. How are you doing that? Are you getting the charger to send StatusNotification messages back to Home Assistant every minute? Sounds like a memory leak in the OCPP code in the firmware, if increasing the OCPP traffic to the charger can get it to fail faster.

S&P do seem interested in fixing these problem. I've not got any complaints with their engagement on this. The last I heard from them was they were going to setup one of their chargers with Home Assistant. And have a console connection to the charger to see if that helps diagnose the problem.  Fingers crossed.

**Reply by bleara** [reply]
Thanks for the info.
Originally I was using automations for everything I do but I am migrating over to node-red. Setting up a flow that reads the status every minute is so straight forward and then notifying me via WhatsApp of the error state very useful.
David Simpson contacted me today to say they are engaging the services of a developer next week who is an expert in HA so hopefully this issue will be resolved quickly. They are keen to get it fixed, as am I.

**Reply by bleara** [reply]
Since the unavailability on the 22nd it has been fine ever since for 5 days despite me increasing the sampling of the status, although we did have a brief power cut 3 days ago which might affect stats. So, I did initially have a couple of failures with 2.2.39 but it now seems much more stable and I can't be sure if anything I do affects the reliability. I see that 'm01' has 2.2.40 but mine still reports 2.2.39. 
On another point: I originally installed the Android App on my phone in December 2023 and I noticed that it was now version 1.9 but the latest version is 1.10.1. Google Play didn't offer an upgrade or even recognise the App as installed so I had to uninstall it in the settings and then re-install it from scratch. I don't know if the same might happen with Apple.

**Reply by edc1** [reply]
Just out of interest, to get the charger status are you frequently sending StatusNotification messages to the charger? Like this:

```
2024-11-27 19:00:38.066 INFO (MainThread) [ocpp] : send [2,"93413b78-f66e-43dc-962a-d0349ef3da2b","TriggerMessage",{"requestedMessage":"StatusNotification","connectorId":0}]
2024-11-27 19:00:38.251 INFO (MainThread) [ocpp] : receive message [3,"93413b78-f66e-43dc-962a-d0349ef3da2b",{"status":"Accepted"}]
2024-11-27 19:00:38.251 INFO (MainThread) [ocpp] : receive message [2,"xxxxxxxxxxxx-066826e0-000001ac","StatusNotification",{"connectorId":0,"status":"Available","errorCode":"NoError"}]
```

**Reply by bleara** [reply]
I just poll the status value "sensor.charger_status" which I expected would have no direct impact on the OCPP as I assumed it would be updated periodically from the OCPP interface by the driver. I was surprised that it did seem to have an effect, my knowledge of the internal workings is still sketchy though. 
When it was failing it didn't seem to make any difference whether the car was connected or not, or charging or not.

**Reply by m01** [reply]
> I see that 'm01' has 2.2.40 but mine still reports 2.2.39.

Note that I observed the version 2.2.40 on a Business 7 charger; maybe a difference between the home & business models' version numbers is normal.

**Reply by bleara** [reply]
I have had the charger go 'unavailable' twice yesterday. The first occasion was for over 2 hours in the middle of a charge so the charge stopped permanently as I hadn't got my error recovery code active. The charger did go into the 'preparing' state after the 2 hours but that would have been a large chunk of the charge time lost even if I had the code active. Charged last night though.
I'm not sure of the difference between 'preparing' and 'available', anyone know?
David Simpson did contact me last night to say they are working on the problem over the weekend.

**Reply by edc1** [reply]
The charger goes from Available to Preparing when you plug the car in. Essentially Preparing means connected to the car but not charging.

Here's what the status looks like in Home Assistant for a successful charge, from the perspective of the charger (Status Connector and Availability) and the car (Charging System Status):

![unnamed](https://github.com/user-attachments/assets/0eaaf496-91f8-4284-a5a1-e453fc32ebc3)

The small brown-ish coloured block is SuspendedEV status where the car had reached it charge target.

Here are the definitions for the various different statuses as defined in the OCPP 1.6 spec:

| STATUS | CONDITION |
| ------ | --------- |
| Available | When a Connector becomes available for a new user (Operative) |
| Preparing | When a Connector becomes no longer available for a new user but there is no ongoing Transaction (yet). Typically a Connector is in preparing state when a user presents a tag, inserts a cable or a vehicle occupies the parking bay (Operative) |
| Charging | When the contactor of a Connector closes, allowing the vehicle to charge (Operative) |
| SuspendedEVSE | When the EV is connected to the EVSE but the EVSE is not offering energy to the EV, e.g. due to a smart charging restriction, local supply power constraints, or as the result of StartTransaction.conf indicating that charging is not allowed etc. (Operative) |
| SuspendedEV | When the EV is connected to the EVSE and the EVSE is offering energy but the EV is not taking any energy. (Operative) |
| Finishing | When a Transaction has stopped at a Connector, but the Connector is not yet available for a new user, e.g. the cable has not been removed or the vehicle has not left the parking bay (Operative) |
| Reserved | When a Connector becomes reserved as a result of a Reserve Now command (Operative) |
| Unavailable | When a Connector becomes unavailable as the result of a Change Availability command or an event upon which the Charge Point transitions to unavailable at its discretion. Upon receipt of a Change Availability command, the status MAY change immediately or the change MAY be scheduled. When scheduled, the Status Notification shall be send when the availability change becomes effective (Inoperative) |
| Faulted | When a Charge Point or connector has reported an error and is not available for energy delivery . (Inoperative). |

**Reply by bleara** [reply]
It's the 'charger status' rather than 'charger status connector' I have been looking at which doesn't matter if it's preparing or available just so long as it isn't unavailable. This is my successful 2 days of charging.
![charging](https://github.com/user-attachments/assets/96e27a79-33dd-4a17-a905-12ab8df3f0a4)
The 'charger status connector' is a more useful status.

**Comment by m01** [comment]
> This is particularly annoying when you need it to charge the car in the middle of the night and it doesn't.

How did you setup the timed charging? Did you set the maximum current to 0A to pause charging until the right time? (like here: https://github.com/lbbrhzn/ocpp/issues/1062#issuecomment-1918007817) If so, did that work?

**Reply by edc1** [reply]
No, I don't do that.

I'm an Octopus Energy customer in the UK on their Agile tariff. I use the Octopus integration in HA. That lets me configure target rate sensors. I have a target rate sensor that comes on when the electricity price meets the criteria I set. And goes off again when the criteria aren't met.

When the target rate sensor comes on, automation (in Node Red) turns the charge_control switch on. That sends a RemoteStartTransaction message over OCPP to the charger, and the car starts charging. Same thing when the target rate sensor goes off. Automation sets charge_control to off and a RemoteStopTransaction message is sent to the charger. 

I also have a charge target set on the car. So if that car gets to 80% it tells the charger to stop.

Of course this doesn't work if the charger is in a Faulted state. Which is what's making this setup unreliable.

**Comment by edc1** [comment]
There are two seperate issues that S&P are looking at for me now. The first is the charger regulary reporting InternalError errors like this:

```
2024-11-25 05:14:30.165 INFO (MainThread) [ocpp] : receive message [2,"xxxxxxxxxxxx-4c598d93-00000029","StatusNotification",{"connectorId":1,"status":"Faulted","errorCode":"InternalError"}]
2024-11-25 05:14:30.168 INFO (MainThread) [ocpp] : send [3,"xxxxxxxxxxxx-4c598d93-00000029",{}]
2024-11-25 05:14:40.402 INFO (MainThread) [ocpp] : receive message [2,"xxxxxxxxxxxx-643bd725-0000002a","StatusNotification",{"connectorId":1,"status":"Available","errorCode":"NoError"}]
2024-11-25 05:14:40.406 INFO (MainThread) [ocpp] : send [3,"xxxxxxxxxxxx-643bd725-0000002a",{}]
```

The second is an issue with the charger not reconnecting to Home Assistant after it restarts. From the websocket server logs:

```
2024-11-03 17:49:56.078 DEBUG (MainThread) [websockets.server] < PING '' [0 bytes]
2024-11-03 17:49:56.079 DEBUG (MainThread) [websockets.server] > PONG '' [0 bytes]
2024-11-03 17:50:06.014 DEBUG (MainThread) [websockets.server] > PING 92 14 d2 a2 [binary, 4 bytes]
2024-11-03 17:50:14.609 DEBUG (MainThread) [websockets.server] = connection is CONNECTING
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < GET / HTTP/1.1
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Connection: Upgrade
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Host: xxx.xxx.xxx:9000
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < User-Agent: ESP32 Websocket Client
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Upgrade: websocket
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Sec-WebSocket-Version: 13
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Sec-WebSocket-Key: /+zS60gG3tJObiTsdQ2lEQ==
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Sec-WebSocket-Protocol: ocpp1.6
2024-11-03 17:50:14.611 DEBUG (MainThread) [websockets.server] < Authorization: Basic NDAyMmQ4ZjFiM2I0OjEyMzQ1Njc4
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > HTTP/1.1 101 Switching Protocols
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Upgrade: websocket
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Connection: Upgrade
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Sec-WebSocket-Accept: SKa1YTL+n4b+fxuAp0xq2hDi9Js=
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Sec-WebSocket-Protocol: ocpp1.6
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Date: Sun, 03 Nov 2024 17:50:14 GMT
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > Server: Python/3.12 websockets/13.1
2024-11-03 17:50:14.612 INFO (MainThread) [websockets.server] connection open
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] = connection is OPEN
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] = connection is CLOSING
2024-11-03 17:50:14.612 DEBUG (MainThread) [websockets.server] > CLOSE 1000 (OK) [2 bytes]
2024-11-03 17:50:14.990 DEBUG (MainThread) [websockets.server] < TEXT '[2,"xxxxxxxxxxxx-5851f42d-00000001","BootNotifi...wareVersion":"2.2.38"}]' [198 bytes]
2024-11-03 17:50:20.915 DEBUG (MainThread) [websockets.server] = connection is CLOSED
2024-11-03 17:50:20.916 DEBUG (MainThread) [websockets.server] ! failing connection with code 1006
2024-11-03 17:50:20.917 DEBUG (MainThread) [websockets.server] x half-closing TCP connection
2024-11-03 17:50:20.920 DEBUG (MainThread) [websockets.server] > TEXT '[3,"xxxxxxxxxxxx-5851f42d-00000001",{"currentTi...0,"status":"Accepted"}]' [111 bytes]
2024-11-03 17:50:20.920 DEBUG (MainThread) [websockets.server] = connection is CLOSING
2024-11-03 17:50:20.920 DEBUG (MainThread) [websockets.server] > CLOSE 1000 (OK) [2 bytes]
2024-11-03 17:50:21.508 DEBUG (MainThread) [websockets.server] < CLOSE 1005 (no status received [internal]) [0 bytes]
2024-11-03 17:50:21.508 DEBUG (MainThread) [websockets.server] x half-closing TCP connection
2024-11-03 17:50:21.731 DEBUG (MainThread) [websockets.server] = connection is CLOSED
2024-11-03 17:50:21.731 INFO (MainThread) [websockets.server] connection closed
2024-11-03 17:50:21.731 INFO (MainThread) [websockets.server] connection closed
```

**Comment by blacknell** [comment]
Did these issues get sorted on 2.2.47?

**Comment by bleara** [comment]
Simpsons have been excellent on supporting our issues and Emlyn and I have been Beta testing for them over the last few months. They have been downloading code to our chargers and we have been reporting back with logs etc when we have seen issues. The version we are testing at the moment is 2.2.50 which seems solid. I am not sure when this will be on general release but I would imagine this shouldn't be long.

**Reply by blacknell** [reply]
Can you share what's new and what's been fixed please

**Reply by blacknell** [reply]
I've had a number of problems getting to grips with my new EV and charger which I discussed in #1551

**Comment by bleara** [comment]
As Emlyn commented in the other thread if you are using Intelligent Octopus Go because your car supports it then you shouldn't be trying to use HA to control it too. 
I can't provide a definitive answer to what has been fixed, you would have to talk to Simpsons for that. 
I don't have IOG support for my car so I purely use HA. I detect the car is connected using the cars status but you could use charger_status_connector == preparing, or charger_availability == Off. Then at 00:30 I then turn on charger_charge_control and turn it off at 05:30. I also check the SOC of the car so I can stop the charge early at a level set in HA rather than the cars set limit. I also have a charge restart routine to restart the charge if it is interrupted within the off peak period for whatever reason.

**Comment by 8muirfield** [comment]
Hi All,

I've just got a S&P Home Series V3 Home Plus charger installed and trying now to link to home assistant. I've followed the instructions for OCPP, but unable to get any entities to show in home assistant and not sure I am putting the right URL into the charger app either. I think the Key points are what format to use in the app to link and also what settings to put into OCPP when configuring. Can anyone advise?

Thanks :-)

**Comment by daviesalex** [comment]
I wanted to share a positive experience with this (from 2026). I'm not actually using HA, but I think my findings will be useful for folks who are or those working on HA integrations. 

I bought a bunch of Simpson & Partners Home 22kW Plus chargers, all physically hard wired via Ethernet. Claude built an OCPP 1.6 central system for me (Python, `ocpp` + `websockets`) that runs on a Rasberry Pi on my local network that controls charging, gating it to cheap/negative Octopus Agile price windows. I use the same system to charge a home battery (via MODBUS commands). I have it all working just fine. 

Sharing a few findings since some of this thread's symptoms match what I hit or was not sure about before I made the purchase:

**1. Same app bug configuring the OCPP server URL, same workaround.** Characters disappeared while typing the backend URL in the iPhone app — pasting the full URL in one go instead of typing worked around it. 

**2. I hit a firmware bug presenting purely as random disconnects.** Firmware above 3.4.0 and below 3.4.19 derives its LAN MAC incorrectly (keeps only the first 4 factory-MAC bytes, replaces the rest with a fixed `ff:fe`). With more than one affected unit on the same network, they all present the identical MAC and fight over one DHCP lease — showed up as ~30-60s reconnect cycles, "paired" dropouts, and the shared IP always answering pings even while a specific charger's session was dead.

While troubleshooting (2) I moved everything to a secure web socket (i'm still not sure if this is necessary or not, but its better) - I set up Let's Encrypt against an internet domain name pointed at an internal IP (`foo.bar.com` → `192.168.x.y`), which worked fine.

**3. Vendor was genuinely and extremely responsive** — once I figured out (with claude's help) the MAC address problem in (2), one support email was forwarded to the CEO, who replied the same day, and new firmware was pushed within the same week. Very, very impressive. The new firmware for sure fixes the bug. 

**4. What our server does is genuinely simple.** On the receiving side it just handles the standard lifecycle messages — `BootNotification`, `Heartbeat`, `Authorize` (we accept every tag for home use), `StartTransaction`/`StopTransaction`, `StatusNotification`, and `MeterValues`. On telemetry specifically: out of the box we get `Energy.Active.Import.Register` periodically with no configuration needed; sending `ChangeConfiguration` for `MeterValueSampleInterval` + `MeterValuesSampledData` (to also get `Power.Active.Import`, i.e. live kW) works cleanly on 3.4.19 too — applies to an already-running transaction, no socket drop. Nothing exotic. On the sending side, it uses exactly two commands: `RemoteStartTransaction` when a connector is waiting for power and our price gate currently says yes, and `RemoteStopTransaction` when the gate closes on a running session. That's the entire mechanism — no `ChangeAvailability`, no charging profiles, no configuration pushed proactively otherwise. Withholding `RemoteStart` is enough to prevent charging outside allowed windows on a fleet with no free-vend.

This has been running unattended for weeks, gating real charging sessions purely off live pricing, with zero user interaction required at the charger or in any vendor app. If you're weighing up whether to build something like this yourselves: it works, and it doesn't need to be complicated to do so.

**5. Tesla-specific behaviour**: the car shows "charging equipment not ready" in the Tesla app initially, but that's fine — `RemoteStartTransaction` → the charger's `StartTransaction`, with no `Authorize` call or any other message from the car in between, and it charges. No user action needed once it's plugged in, including in the middle of the night. It's also able to stop and start again cleanly if a charge needs more than one price window to complete.
